cockroachdb in comparison

jseldess opened the pull request:
@bdarnell, @mrtracy, @spencerkimball, @petermattis, anyone else who wants to help:

I'm aiming to have a quick yes/no feature comparison in the beta docs. At this stage, I'd appreciate help identifying the right "features" for such a quick comparison.

- Did I list the right features? 
- Should any be rephrased?
- Are any missing?
- Any other thoughts?

Here's an html version (still needs styling), in case it's easier to read: http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/cockroachdb-in-comparison.html

I'll find a way to provide a bit more detail (sentence or two) on hover over a yes/no answer. Hopefully we can link to standalone pages (or page sections) in our docs for each feature.

spencerkimball commented on cockroachdb-in-comparison.md:
```diff
@@ -2,3 +2,20 @@
 toc: false
 ---
+
+| | CockroachDB | MySQL | Postgre | Cassandra | HBase | MongoDB | Riak 
```
s/Postgre/PostgreSQL/

jseldess replied:
Fixing now.

spencerkimball commented on cockroachdb-in-comparison.md:
```diff
@@ -2,3 +2,20 @@
+| | CockroachDB | MySQL | Postgre | Cassandra | HBase | MongoDB | Riak 
+--|--|--|--|--|--|--|--
+Automated Scaling & Repair| Yes | No | No |  |  | No | Yes 
+Synchronous Replication | Yes | No | No |  |  | No |  
```
I'd drop this one – it's largely the same thing as "strong consistency"

jseldess replied:
OK, I'll move Strong Consistency up here. Is it necessary to get the term "replication" in somewhere, though?

spencerkimball commented on cockroachdb-in-comparison.md:
```diff
@@ -2,3 +2,20 @@
+Automated Scaling & Repair| Yes | No | No |  |  | No | Yes 
+Synchronous Replication | Yes | No | No |  |  | No |  
+Automatic Failover | Yes | No | No |  |  |  | Yes 
+Read from Any Replica | No |  |  |  |  | Yes |  
```
CockroachDB does have "Read from any Replica" but it's eventually consistent.

Here's a link to my estimation of the comparison matrix: https://docs.google.com/a/cockroachlabs.com/spreadsheets/d/1ZHhLkku2CJkUhyoAT_NUqOQmAPDC2cJ_2mfK4NXcarw/edit?usp=sharing

jseldess replied:
Thanks for those estimations. I'll put them in now.

For "Read from any Replica", @bdarnell mentioned to me that it's possible at the kv layer but not through sql. Is that what you mean here? If so, since we're not encouraging kv access (is it even possible?), shouldn't we mark this "No", maybe with a caveat that it'll be possible later?

jseldess replied:
@spencerkimball: One question on Riak. In your estimation, you say "No" for strong consistency, but they're docs say "Yes": http://docs.basho.com/riak/latest/dev/advanced/strong-consistency/. Thoughts?

spencerkimball commented:
CockroachDB does provide "read from any replica" btw. It's eventually consistent as with the other distributed DBs though.

jseldess commented:
@spencerkimball: Sorry about that. I'll change read from any replica to yes for CockroachDB. But one of my earlier questions got lost: How do users do this? Ben mentioned this isn't possible through SQL. Do we give users access to the KV layer right now?

vivekmenezes commented on cockroachdb-in-comparison.md:
```diff
@@ -2,3 +2,19 @@
+| | CockroachDB | MySQL | PostgreSQL | Cassandra | HBase | MongoDB | Riak 
+--|--|--|--|--|--|--|--
+Automated Scaling | Yes | No | No | Yes | Yes  | Yes | Yes 
+Automated Failover & Repair | Yes | No | No | Yes | Yes | Yes | Yes 
```
I'd just call this Automated Failover. The repair suggests that it is consistent.

jseldess replied:
Thanks, @vivekmenezes. I'll make that change.

spencerkimball replied:
It isn't currently possible through SQL, though that would be relatively easy to change. For now we can go ahead and say "No" for that feature I suppose.

vivekmenezes replied:
It might make sense to reword it as

- eventual consistent read from any replica

to which our answer would be no for now.


On Wed, Feb 17, 2016 at 11:08 AM Spencer Kimball <notifications@github.com>
wrote:

> In cockroachdb-in-comparison.md
> <https://github.com/cockroachdb/docs/pull/52#discussion_r53186478>:
>
> > @@ -2,3 +2,19 @@
> >  title: CockroachDB in Comparison
> >  toc: false
> >  ---
> > +
> > +| | CockroachDB | MySQL | PostgreSQL | Cassandra | HBase | MongoDB | Riak
> > +--|--|--|--|--|--|--|--
> > +Automated Scaling | Yes | No | No | Yes | Yes  | Yes | Yes
> > +Automated Failover & Repair | Yes | No | No | Yes | Yes | Yes | Yes
>
> It isn't currently possible through SQL, though that would be relatively
> easy to change. For now we can go ahead and say "No" for that feature I
> suppose.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/docs/pull/52/files#r53186478>.
>

bdarnell commented on cockroachdb-in-comparison.md:
```diff
@@ -2,3 +2,19 @@
+
+Need to find a way to provide addition detail on hover:
+
+\* Apache Pheonix can provide a SQL layer on top 
```
Typo: Phoenix

And it's [not the only one](http://www.infoworld.com/article/2683729/hadoop/10-ways-to-query-hadoop-with-sql.html). It's not even the only one under the apache umbrella: [trafodion](http://trafodion.apache.org/). I think it would be best to only consider plain hbase here, and if we want to talk about a sql layer on top of hbase then it should get its own column (since some of the sql layers may have different answers for some of the other questions)

jseldess replied:
Makes sense, Ben. I'll take that out.

jseldess commented:
- Styled the table to resemble those in our blog posts. See here: http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/cockroachdb-in-comparison.html
- Added an intro sentence emphasizing that the page compares key cockroachdb features against other databases. Given that focus and the back and forth about reading from any replica, I just removed that row. @spencerkimball, @bdarnell, @vivekmenezes, that ok?
- For each feature, linked to related FAQ.

Open questions: 

- Should we add any other features, such as "Lock-Free Transactions"?
- I'll check the other dbs' docs to see if we can support our answers, but does anyone disagree with any answers based on personal experience with these databases?

vivekmenezes commented:
- can you fit each feature on a single line.
- Removing reads from any replica makes sense because the other databases
support this as "inconsistent reads from any replica"
- Should we rename "strong consistency" as "consistent replication"

On Wed, Feb 17, 2016 at 11:41 PM jseldess <notifications@github.com> wrote:

>
>    - Styled the table to resemble those in our blog posts. See here:
>    http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/cockroachdb-in-comparison.html
>    - Added an intro sentence emphasizing that the page compares key
>    cockroachdb features against other databases. Given that focus and the back
>    and forth about reading from any replica, I just removed that row.
>    @spencerkimball <https://github.com/spencerkimball>, @bdarnell
>    <https://github.com/bdarnell>, @vivekmenezes
>    <https://github.com/vivekmenezes>, that ok?
>    - For each feature, linked to related FAQ.
>
> Open questions:
>
>    - Should we add any other features, such as "Lock-Free Transactions"?
>    - I'll check the other dbs' docs to see if we can support our answers,
>    but does anyone disagree with any answers based on personal experience with
>    these databases?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/docs/pull/52#issuecomment-185540587>.
>

jess-edwards commented:
I think we should add at least one if not two closed source products to this list - AWS Aurora / RDS and Microsoft SQL make the most sense to me. 

I also think this table looks suspicious with 100% "yes" in the CockroachDB column. I think you should add the following rows: 
+ read from any replica
+ geo-replicated (Is this possible currently or should we add a caveat like "Not yet")
+ enterprise version (Not yet)
+ support (Limited) 

Re: Riak - Strongly consistent: Should be "limited" or "sometimes" or "optional."

Move the hyperlinks from the features to each of the yes/nos. @jseldess I will help you find supporting links for each DB.

jseldess commented:
Latest commit:
- Removed Riak, added Oracle and DynamoDB.
- Added tooltip for each feature (e.g., `Automated Scaling`) to state intended meaning.
- Put back in `Eventually Consistent Reads`, and added `Commercial Version` and `Support`.
- Added `Automated Repair`, since some dbs may be able to failover automatically but not repair missing data.

I've tried tracking down supporting docs for some answers, but it's turning into a rabbit hole and might be overkill for beta. Instead, I'm going to set up a quick meeting with @spencerkimball, @bdarnell, and @jess-edwards for next week to work out final(ish) answers.

jseldess commented on js/customscripts.js:
```diff
@@ -71,20 +71,20 @@
 
 
         //hijack clicks on links, verify if anchor link and jump to div
-        $('body').on('click', 'a', function (e) {
```
@mjibson, commented this bit out to fix that anchor issue.

jseldess commented on js/customscripts.js:
```diff
@@ -71,20 +71,20 @@
 
 
         //hijack clicks on links, verify if anchor link and jump to div
```
@mjibson, commented the following bit out to fix that anchor issue.

mjibson replied:
Any reason not to remove it altogether?

jseldess replied:
Probably not. Just didn't have time to see what else in the theme might be depending on it. I'll test around and make sure everything's fine and delete this, if so.

jseldess replied:
Nope. Removing.

mjibson commented on css/theme-blue.css:
```diff
@@ -1,3 +1,28 @@
+    border-left: 2px solid #FFFFFF;
+    background: #e8e7e5;
+    color: #142848;
+    font-family: 'Inconsolata';
```
A font family here looks pretty fishy to me. Why is this neeeded?

jseldess replied:
That's there so our tables match table formatting on the blog. But I agree that it's not necessary here, especially since this content shouldn't be monospaced. I'll remove.

BramGruneir commented:
I agree with @jess-edwards that we should include some closed source alternatives such.

jseldess commented:
@spencerkimball, @bdarnell, @jess-edwards: Updated page based on our discussion this morning. HTML version here: http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/cockroachdb-in-comparison.html

Anything else to change?

bdarnell commented:
LGTM

vivekmenezes commented:
LGTM

On Mon, Feb 22, 2016 at 6:39 PM Ben Darnell <notifications@github.com>
wrote:

> LGTM
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/docs/pull/52#issuecomment-187440045>.
>
